### PR TITLE
rewrite to use ann query operator and recent research

### PR DESCRIPTION
**Pull request by jobergum**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: documentation/semantic-qa-retrieval.html [hunk 1]
```diff
@@ -1,66 +1,93 @@
...
+documents from the first stage.
 </p>
 
 <p>
-In <a href="https://arxiv.org/abs/1907.04780">ReQA: An Evaluation for End-to-End Answer Retrieval Models</a> Ahmad Et al. introduce Retrieval Question Answering (ReQA),a benchmark 
-for evaluating large-scale sentence level answer retrieval models and where they establish a baseline for both traditional information retrieval techniques and neural encoding models on the 
-<em>Stanford Question Answering Dataset (SQuAD)</em> v1.1 dataset. 
-In this document we reproduce the work done by Ahmad et al. on the <a href="https://rajpurkar.github.io/SQuAD-explorer/">SQuAD 1.1</a> retrieval task using Vespa serving engine. 
+In <a href="https://arxiv.org/abs/1907.04780">ReQA: An Evaluation for End-to-End Answer Retrieval Models</a> Ahmad Et al. introduce 
+<em>Retrieval Question Answering (ReQA)</em>,a benchmark
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:31]:
```suggestion
<em>Retrieval Question Answering (ReQA)</em>, a benchmark 
```

File: documentation/semantic-qa-retrieval.html [hunk 2]
```diff
@@ -1,66 +1,93 @@
...
+organization to deploy state of the art question answering retrieval systems with low effort using the scalable Vespa engine.</p>
 
 <p>
-Vespa has support for storing tensors as fields in the Vespa document model which we can performing operations over tensors fields in addition to traditional field types and ranking features 
-like <a href="reference/bm25.html">bm25</a> or Vespa's <a href="reference/nativerank.html">nativeRank</a>.  Having both traditional text ranking features and semantic similarity features
-expressed in the same engine is a powerful feature of Vespa.
- 
-The work described in this document can be reproduced using the <a href="https://github.com/vespa-engine/sample-apps/tree/master/semantic-qa-retrieval">semantic-qa-retrieval sample application</a>. 
+Vespa has support for storing and indexing dense tensors field types along with traditional 
+string fields with support for traditional sparse term based text ranking features .
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:44]:
```suggestion
string fields with support for traditional sparse term based text ranking features
```

File: documentation/semantic-qa-retrieval.html [hunk 3]
```diff
@@ -1,66 +1,93 @@
...
-Vespa has support for storing tensors as fields in the Vespa document model which we can performing operations over tensors fields in addition to traditional field types and ranking features 
-like <a href="reference/bm25.html">bm25</a> or Vespa's <a href="reference/nativerank.html">nativeRank</a>.  Having both traditional text ranking features and semantic similarity features
-expressed in the same engine is a powerful feature of Vespa.
- 
-The work described in this document can be reproduced using the <a href="https://github.com/vespa-engine/sample-apps/tree/master/semantic-qa-retrieval">semantic-qa-retrieval sample application</a>. 
+Vespa has support for storing and indexing dense tensors field types along with traditional 
+string fields with support for traditional sparse term based text ranking features .  
+like <a href="reference/bm25.html">bm25</a> or Vespa's <a href="reference/nativerank.html">nativeRank</a>.  
+Having both traditional text ranking features and semantic similarity features
+expressed in the same engine is a powerful feature of Vespa which enables hybrid retrieval using both sparse and dense representation.</p>
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:47]:
```suggestion
expressed in the same engine is a powerful feature of Vespa, which enables hybrid retrieval using both sparse and dense representation.</p>
```

File: documentation/semantic-qa-retrieval.html [hunk 4]
```diff
@@ -1,66 +1,93 @@
...
+The <em>Universal Sentence Encoder</em> encodes text into fixed length dense embedding space that can be used for broad range of tasks such as semantic similarity, 
+ semantic retrieval and other natural language processing (NLP) tasks.
 Google has released several different sentence encoder models with different goals and following the work of Ahmad et al. we use the 
 <a href="https://ai.googleblog.com/2019/07/multilingual-universal-sentence-encoder.html">Multilingual Universal Sentence Encoder for Question-Answer Retrieval</a>.
 
-The Universal Sentence Encoder for Question-Answer Retrieval enables us to process questions and candidate answer sentences independently and map the high dimensional sparse text representation to a relatively low dimensional
-dense tensor representation. 
+The Universal Sentence Encoder for Question-Answer Retrieval enables us to process questions and candidate answer sentences independently and map the 
+high dimensional sparse text representation to a relatively low dimensional
+dense tensor representation where we can use Vespa's approximate nearest neighbor search operator to retrieve documents efficiently.
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:63]:
```suggestion
dense tensor representation, where we can use Vespa's approximate nearest neighbor search operator to retrieve documents efficiently.  
```

File: documentation/semantic-qa-retrieval.html [hunk 5]
```diff
@@ -1,66 +1,93 @@
...
-dense tensor representation. 
+The Universal Sentence Encoder for Question-Answer Retrieval enables us to process questions and candidate answer sentences independently and map the 
+high dimensional sparse text representation to a relatively low dimensional
+dense tensor representation where we can use Vespa's approximate nearest neighbor search operator to retrieve documents efficiently.  
 
 <ul>
-<li>Question text is encoded using the question encoder which takes the question text as input and outputs a 512 dimensional dense tensor. This can be done at user time in a couple of ms</li>
-<li>Each sentence of text (Which might contain the answer to a given question) is encoded using the response encoder which takes the sentence and the surrounding context (e.g paragraph level) as input and outputs a 512 dimension dense tensor</li> 
+<li>Question text is encoded using the question encoder which takes the question text as input and outputs a 512 dimensional dense tensor.</li>
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:66]:
```suggestion
<li>Question text is encoded using the question encoder, which takes the question text as input and outputs a 512 dimensional dense tensor.</li> 
```

File: documentation/semantic-qa-retrieval.html [hunk 6]
```diff
@@ -1,66 +1,93 @@
...
+The Universal Sentence Encoder for Question-Answer Retrieval enables us to process questions and candidate answer sentences independently and map the 
+high dimensional sparse text representation to a relatively low dimensional
+dense tensor representation where we can use Vespa's approximate nearest neighbor search operator to retrieve documents efficiently.  
 
 <ul>
-<li>Question text is encoded using the question encoder which takes the question text as input and outputs a 512 dimensional dense tensor. This can be done at user time in a couple of ms</li>
-<li>Each sentence of text (Which might contain the answer to a given question) is encoded using the response encoder which takes the sentence and the surrounding context (e.g paragraph level) as input and outputs a 512 dimension dense tensor</li> 
+<li>Question text is encoded using the question encoder which takes the question text as input and outputs a 512 dimensional dense tensor.</li> 
+<li>Each sentence of text is encoded using the 
+response encoder which takes the sentence and the surrounding context (e.g paragraph level) as input and outputs a 512 dimension dense tensor</li>
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:68]:
```suggestion
response encoder, which takes the sentence and the surrounding context (e.g paragraph level) as input and outputs a 512 dimension dense tensor</li> 
```

File: documentation/semantic-qa-retrieval.html [hunk 7]
```diff
@@ -1,66 +1,93 @@
...
 <p>Papers and resources on Google's Universal Sentence Encoder:</p>
 <ul>
   <li><a href="https://arxiv.org/abs/1907.04307">Multilingual Universal Sentence Encoder for Semantic Retrieval Paper</a> </li>
   <li><a href="https://ai.googleblog.com/2019/07/multilingual-universal-sentence-encoder.html">Google AI Blog: Multilingual Universal Sentence Encoder for Semantic Retrieval</a> </li>
   <li><a href="https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/1">Tensorflow Hub: USE-QA </a> </li>
 </ul>
-<p>
+
+<p>A similar dual encoder architecture (question, document) is described in the <a href="https://arxiv.org/abs/2004.04906">Dense Passage Retrieval for Open-Domain Question Answering</a>
+by Facebook Research where they demonstrate how a trained dense representation using a dual question encoder based on BERT outperforms traditional IR retrieval. Quote:
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:87]:
```suggestion
by Facebook Research, where they demonstrate how a trained dense representation using a dual question encoder based on BERT outperforms traditional IR retrieval. Quote: 
```

File: documentation/semantic-qa-retrieval.html [hunk 8]
```diff
@@ -139,38 +168,31 @@ <h2 id="vespa-data-model">SQuAD Data modelling with Vespa</h2>
...
-The Vespa tensor field type cannot be searched as regular <a href="reference/schema-reference.html#field-types">field types</a> but can be
-used for ranking. There exist techniques which can be used in combination with tensor fields in Vespa 
-which can enable search or approximate nearest neigboor search over other data structures than 
-the tensor fields, e.g over weighted sets in Vespa 
-using <a href="https://en.wikipedia.org/wiki/Locality-sensitive_hashing">Locality Sensitive Hashing</a> for dimension reduction which 
-enables sparse representations which can be efficiently evaluated using e.g the Vespa <a href="using-wand-with-vespa.html">WAND</a> implementation. 
-Such approximation techniques are though out of scope for this particular document.
+<p>
+See <a href="approximate-nn-hnsw.html">Approximate Nearest Neighbor Search using HNSW Index</a> for details on the <em>HNSW</em> index settings and <em>distance-metric</em>. 
+In our case we use the <em>euclidean</em> distance metric.
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:186]:
```suggestion
In this case, we use the <em>euclidean</em> distance metric. 
```

File: documentation/semantic-qa-retrieval.html [hunk 9]
```diff
@@ -190,12 +212,12 @@ <h2 id="vespa-feed-generating">Converting the SQuAD json to Vespa json feed form
     }
 }
 </pre>
-We can feed the generated document set to Vespa using any of the feed api's but we use the <a href="vespa-http-client.html">Vespa http client</a>. After this step we have one content db
+<p>
+We can feed the generated document set to our Vespa instance using any of the feed api's but
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:216]:
```suggestion
We can feed the generated document set to our Vespa instance using any of the feed api's, but 
```

File: documentation/semantic-qa-retrieval.html [hunk 10]
```diff
@@ -190,12 +212,12 @@ <h2 id="vespa-feed-generating">Converting the SQuAD json to Vespa json feed form
...
-We can feed the generated document set to Vespa using any of the feed api's but we use the <a href="vespa-http-client.html">Vespa http client</a>. After this step we have one content db
+<p>
+We can feed the generated document set to our Vespa instance using any of the feed api's but 
+we use the <a href="vespa-http-client.html">Vespa http client</a>. After this step we have one content db
 with 18,896 context documents and 91,729 sentences in another in the same Vespa content cluster. 
 
-
-
-<h2 id="sentence-retrieval">Sentence Retrieval</h2>
+<h2 id="sentence-paragraph-retrieval">Sentence and Paragraph Retrieval</h2>
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:220]:
```suggestion
<h2 id="sentence-and-paragraph-retrieval">Sentence and Paragraph Retrieval</h2>
```

File: documentation/semantic-qa-retrieval.html [hunk 11]
```diff
@@ -204,83 +226,72 @@ <h2 id="sentence-retrieval">Sentence Retrieval</h2>
...
+<pre>
+{
+  'yql': 'select * from sources sentence where ([{"targetNumHits":100}]nearestNeighbor(sentence_embedding,query_embedding))| \\
+    all(group(context_id) max(100) order(-max(relevance())) each( max(2) each(output(summary())) as(sentences)) as(paragraphs));',
+  'hits': 0,
+  'ranking.features.query(query_embedding)': [-0.0466, ...,0.064],
+  'ranking.profile': 'sentence-semantic-similarity' 
+}
+</pre>
+<p>The grouping expression groups sentences by the context id and order the groups ( paragraphs ) by the maximum rank score and for each unique context id we get the top ranking sentences sentences ordered
```
Comment (kkraune) [inline documentation/semantic-qa-retrieval.html:273]:
```suggestion
<p>The grouping expression groups sentences by the context id and order the groups ( paragraphs ) by the maximum rank score. For each unique context id, we get the top ranking sentences sentences ordered
```

**Review by kkraune — APPROVED** [review]
This looks great! I will re-read the doc after merging, lots of changes here :-)

**Comment by jobergum** [comment]
Thanks for the deep and helpful review @kkraune !

**Pull request merged**